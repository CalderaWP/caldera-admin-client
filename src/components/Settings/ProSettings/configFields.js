import {optionFactory} from "../../util/optionFactory";

export const PUBLIC_KEY = 'proPublicKey';
export const PRIVATE_KEY = 'propPrivateKey';
export const ENHANCED_DELIVERY = 'enhancedDelivery';
export const LOG_LEVEL = 'logLevel';
export const GENEREATE_PDFS = 'generatePDFs';

export default {
	apiKeys: [
		{
			id: PUBLIC_KEY,
			label: 'Public Key',
			desc: 'Public Key For Caldera Forms Pro',
			type: 'text',
			path: 'apiKeys.public',
		},
		{
			id: PRIVATE_KEY,
			label: 'Private Key',
			desc: 'Private Key For Caldera Forms Pro',
			path: 'apiKeys.secret',
			type: 'text',
			options: [
				{
					label: 'yes',
					value: 'on'
				}
			]
		},
	],
	generalSettings: [
		{
			id: ENHANCED_DELIVERY,
			label: 'Enable Enhanced Delivery',
			desc: 'If enabled, messages are sent via Caldera Forms Pro Servers',
			type: 'checkbox',
			path: 'generalSettings.enhancedDelivery',
		},
		{
			id: LOG_LEVEL,
			label: 'Minimum Log Level',
			desc: 'Setting a higher level than notice may affect performance, and should only be used when instructed by support.',
			type: 'select',
			path: 'generalSettings.logLevel',
			options: [
				optionFactory(
					100,
					'DEBUG'
				),
				optionFactory(
					200,
					'INFO'
				),
				optionFactory(
					300,
					'WARNING'
				),
				optionFactory(
					400,
					'ERROR'
				),
				optionFactory(
					500,
					'ALERT'
				),
				optionFactory(
					600,
					'EMERGENCY'
				),
			]
		},
	],
	formSettings: [

	]
};





