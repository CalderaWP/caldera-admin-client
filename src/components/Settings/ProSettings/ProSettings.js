import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames'
import {TabPanel} from '@wordpress/components';
import configFields from './configFields'
import {RenderGroup} from '@caldera-labs/components'
import {pick,object} from 'dot-object'
import {proLocalSettingsFactory} from "../factories";

/**
 * Create the CF Pro settings UI
 * @param {Object} props
 * @return {*}
 * @constructor
 */
export class ProSettings extends React.PureComponent {

	/**
	 * Create ProSettings component
	 * @param props
	 */
	constructor(props) {
		super(props);
		this.state = {
			proSettings: proLocalSettingsFactory(props.proSettings)
		};
		this.onSettingsChange = this.onSettingsChange.bind(this);
		this.getConfigFields = this.getConfigFields.bind(this);
		this.onSettingsSave = this.onSettingsSave.bind(this);
		this.tabContentAreaConnected = this.tabContentAreaConnected.bind(this);
	}

	/**
	 * Save the settings
	 */
	onSettingsSave() {
		this.props.onSettingsSave(this.state.proSettings);
	}

	/**
	 * Update state when settings change
	 *
	 * @param {Object} proSettings
	 */
	onSettingsChange(proSettings) {
		this.setState({proSettings});
	}

	/**
	 * Prepare config fields, based on current tab
	 *
	 * @param {String} tabName
	 * @return {Array}
	 */
	getConfigFields(tabName) {
		let currentConfigFields = this.props.configFields[tabName];
		currentConfigFields.forEach(configField => {
			const {path} = configField;
			configField.value = pick(
				path,
				this.state.proSettings
			);

			configField.onValueChange = (newValue) => {
				const update = {
					...this.state.proSettings,
					[path]:newValue
				};
				this.onSettingsChange(object(update));
			};

		});
		return currentConfigFields;
	};

	/**
	 * Renderer for tab content area used when CF Pro is connected
	 * @param tabName
	 * @return {*}
	 */
	tabContentAreaConnected(tabName) {
		return (
			<RenderGroup
				configFields={this.getConfigFields(tabName)}
			/>
		);
	}

	/**
	 * Render the ProSettings UI
	 * @return {*}
	 */
	render() {
		const {connected} = this.state.proSettings;
		let tabs = [

			{
				name: 'apiKeys',
				title: 'Api Keys',
				className: 'pro-settings',
			},
			{
				name: 'generalSettings',
				title: 'Caldera Forms Pro Settings',
				className: 'pro-settings',
			},
			{
				name: 'formSettings',
				title: 'Form Settings',
				className: 'pro-form-settings',
			},
		];
		if( ! connected ){
			tabs = [

				{
					name: 'apiKeys',
					title: 'Api Keys',
					className: 'pro-settings',
				},
				{
					name: 'generalSettings',
					title: 'Caldera Forms Pro Settings',
					className: 'pro-settings',
				},
				{
					name: 'formSettings',
					title: 'Form Settings',
					className: 'pro-form-settings',
				},
			]
		}

		return (
			<TabPanel
				className={classNames(this.props.className, ProSettings.classNames.wrapper)}
				orientation={'vertical'}
				tabs={tabs}
			>
				{
					(tabName) => {
						if( connected ){
							return this.tabContentAreaConnected(tabName);

						}
					}
				}
			</TabPanel>
		)
	}


};

/**
 * Prop types for the ProSettings settings component
 * @type {{}}
 */
ProSettings.propTypes = {
	classNames: PropTypes.string,
	proSettings: PropTypes.object,
	onSettingsSave: PropTypes.func.isRequired,
	configFields: PropTypes.object
};

/**
 * Default props for the ProSettings settings component
 * @type {{}}
 */
ProSettings.defaultProps = {
	configFields
};

/**
 * Class names used in the ProSettings settings component
 * @type {{wrapper: string}}
 */
ProSettings.classNames = {
	wrapper: 'caldera-forms-pro-settings'
};