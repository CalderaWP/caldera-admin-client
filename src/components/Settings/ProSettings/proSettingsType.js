import {
	ENHANCED_DELIVERY,
	LOG_LEVEL,
	PRIVATE_KEY,
	PRO_API_KEYS,
	PRO_GENERAL_SETTINGS,
	PUBLIC_KEY
} from "./configFields";
import PropTypes from 'prop-types'
import {pickArray} from "../../util/pickArray";
import configFields from './configFields';
export const PRO_CONNECTED = 'connected';
export const PRO_SETTINGS = 'proSettings';


export default {
	[PRO_CONNECTED]: PropTypes.bool,
	[PRO_API_KEYS]: PropTypes.shape(
		{
			[PUBLIC_KEY]: PropTypes.string,
			[PRIVATE_KEY]: PropTypes.string,
		}
	),
	[PRO_GENERAL_SETTINGS]: PropTypes.shape(
		{
			[ENHANCED_DELIVERY]: PropTypes.bool,
			[LOG_LEVEL]: pickArray(
				configFields[PRO_GENERAL_SETTINGS].find(setting => LOG_LEVEL === setting.id ).options,
				'value'
			)
		}
	),
	formSettings: PropTypes.object
};