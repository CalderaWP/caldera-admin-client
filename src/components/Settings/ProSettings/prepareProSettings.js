import {settingsProp} from "../../../state/prepareSettings";
import configFields, {
	ENHANCED_DELIVERY,
	LOG_LEVEL,
	PRIVATE_KEY,
	PRO_API_KEYS,
	PRO_GENERAL_SETTINGS,
	PUBLIC_KEY
} from './configFields';
import {PRO_CONNECTED} from './proSettingsType'
import merge from 'deepmerge'

export const prepareProSettings = (settings = {}) => {
	const settingsBasic = {
			[PRO_CONNECTED]: false,
			[PRO_API_KEYS]: {},
			[PRO_GENERAL_SETTINGS]: {}
		};
	if( undefined !== typeof  settings ) {
		settings = merge(
			settingsBasic,
			settings
		);

	}else{
		settings = settingsBasic;
	}


	return {
		[PRO_CONNECTED]: settingsProp(settings, PRO_CONNECTED, false ),
		[PRO_API_KEYS]: {
			[PUBLIC_KEY]: settingsProp(
				settings[PRO_API_KEYS],
				PUBLIC_KEY,
				configFields[PRO_API_KEYS].find( field => PUBLIC_KEY === field.id).default
			),
			[PRIVATE_KEY]: settingsProp(
				settings[PRO_API_KEYS],
				PRIVATE_KEY,
				configFields[PRO_API_KEYS].find( field => PRIVATE_KEY === field.id).default
			),
		},
		[PRO_GENERAL_SETTINGS]: {
			[ENHANCED_DELIVERY]: settingsProp(
				settings[PRO_GENERAL_SETTINGS],
				ENHANCED_DELIVERY,
				configFields[PRO_GENERAL_SETTINGS].find( field => ENHANCED_DELIVERY === field.id).default
			),
			[LOG_LEVEL]: settingsProp(
				settings[PRO_GENERAL_SETTINGS],
				LOG_LEVEL,
				configFields[PRO_GENERAL_SETTINGS].find( field => LOG_LEVEL === field.id).default
			),
		},
		forms: settingsProp(settings,'forms',{})
	};



};
