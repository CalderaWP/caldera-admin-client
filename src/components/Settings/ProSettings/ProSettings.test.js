import renderer from 'react-test-renderer';
import React from 'react';
import {shallow} from 'enzyme';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {ProSettings} from "./ProSettings";

Enzyme.configure({adapter: new Adapter()});
const handler = () => {
};
describe('ProSettings component', () => {

	it( 'Fills in missing settings when mergeing props and state', () => {
		const component = shallow(
			<ProSettings
				onSettingsSave={handler}
			/>);
		expect(
			component.state( 'proSettings' ).hasOwnProperty('connected')
		).toBe(true);
		expect(
			component.state( 'proSettings' ).hasOwnProperty('apiKeys')
		).toBe(true);
		expect(
			component.state( 'proSettings' ).hasOwnProperty('generalSettings')
		).toBe(true);
	});

	it( 'Merges props and state', () => {
		const component = shallow(
			<ProSettings
				proSettings={{
					connected: true,
				}}
				onSettingsSave={handler}
			/>);
		expect(
			component.state( 'proSettings' ).hasOwnProperty('apiKeys')
		).toBe(true);
		expect(
			component.state( 'proSettings' ).hasOwnProperty('connected')
		).toBe(true);

		expect(
			component.state( 'proSettings' ).connected
		).toBe(true);

	});


	it('Matches snapshot with minimal props', () => {
		expect(
			renderer.create(
				<ProSettings
					proSettings={{
						connected: true,
					}}
					onSettingsSave={handler}
				/>
			).toJSON()
		).toMatchSnapshot()
	});

	it('Is wrapped in the right class', () => {
		expect(
			shallow(
				<ProSettings
					proSettings={{
						connected: true,
					}}
					onSettingsSave={handler}
				/>
			).find('.' + ProSettings.classNames.wrapper)
				.length
		).toEqual(1)
	});

	describe('configField preparation', () => {
		describe('apiKeys tab', () => {
			it('has 2 fields', () => {
				const component = shallow(
					<ProSettings
						onSettingsSave={handler}
					/>
				);
				const configFields = component.instance().getConfigFields('apiKeys');
				expect(configFields.length).toEqual(2);
			});

			it('adds value', () => {
				const settings = {
					connected: true,
					apiKeys: {
						public: 'addsValue'
					},
				};
				const component = shallow(
					<ProSettings
						proSettings={settings}
						onSettingsSave={handler}
					/>
				);
				const configFields = component.instance().getConfigFields('apiKeys');
				expect(configFields[0].value).toEqual('addsValue');
			});

			it('adds change handler', () => {
				const component = shallow(
					<ProSettings
						proSettings={
							{
								connected: true
							}
						}
						onSettingsSave={handler}
					/>
				);
				const configFields = component.instance().getConfigFields('apiKeys');
				expect(typeof configFields[0].onValueChange).toEqual('function');
			});

			it('updates state', () => {
				const component = shallow(
					<ProSettings
						proSettings={{
							connected: true,
							apiKeys: {
								public: '',
								private: 'r3s'
							}
						}}
						onSettingsSave={handler}
					/>
				);
				component.instance().onSettingsChange({
					apiKeys: {
						public: 'u'
					}
				});
				expect(component.state('proSettings').apiKeys.public).toEqual('u');
			});
			it('Change handler updates state', () => {
				const component = shallow(
					<ProSettings
						proSettings={{
							connected: true,
							apiKeys: {
								public: '',
								private: 'r3s'
							}
						}}
						onSettingsSave={handler}
					/>
				);
				const configFields = component.instance().getConfigFields('apiKeys');
				configFields[0].onValueChange('updatedValue');
				expect(component.state('proSettings').apiKeys.public).toEqual('updatedValue');
			});

			it('Change handler does not remove state values not updated', () => {
				const component = shallow(
					<ProSettings
						onSettingsSave={handler}
						proSettings={{
							connected: true,
							apiKeys: {
								public: '',
								private: 'r3s'
							}
						}}
					/>
				);
				const configFields = component.instance().getConfigFields('apiKeys');
				configFields[0].onValueChange('updatedValue');
				expect(component.state('proSettings').apiKeys.private).toEqual('r3s');
			});

			it('Passes updates to state to save handle', () => {
				const updatedSettings = {
					proSettings: {
						apiKeys: {
							public: 'hi',
							private: 'roy'
						}
					}
				};

				let updatedSettingsRecived = {};
				const component = shallow(
					<ProSettings
						onSettingsSave={(values) => {
							updatedSettingsRecived = values;
						}}
						proSettings={{
							generalSettings: {
								enhancedDelivery: true,
							},
							apiKeys: {
								public: '',
								private: 'r3s'
							}
						}}
					/>
				);
				component.setState(updatedSettings);
				component.instance().onSettingsSave();
				expect(updatedSettingsRecived.apiKeys.public).toEqual('hi');
				expect(updatedSettingsRecived.apiKeys.private).toEqual('roy');
			});

		});
	});
	describe( 'tab content', () => {
		it( 'Shows apiKeys tab', () => {
			const component = shallow(
				<ProSettings
					onSettingsSave={handler}
					proSettings={{
						connected: true,
						apiKeys: {
							public: '',
							private: ''
						}
					}}
				/>
			);
			const tabArea = renderer.create(
				component.instance().tabContentAreaConnected('apiKeys')
			);
			expect(tabArea.toJSON()).toMatchSnapshot();
		});

		it( 'Shows apiKeys tab', () => {
			const component = shallow(
				<ProSettings
					onSettingsSave={handler}
					proSettings={{
						connected: true,
						apiKeys: {
							public: '',
							private: ''
						}
					}}
				/>
			);
			const tabArea = renderer.create(
				component.instance().tabContentAreaConnected('apiKeys')
			);
			expect(tabArea.toJSON()).toMatchSnapshot();
		});

		it( 'Shows generalSettings tab', () => {
			const component = shallow(
				<ProSettings
					onSettingsSave={handler}
					proSettings={{
						connected: true,
						generalSettings: {
							enhancedDelivery: true,
							generatePDFs: true,
							logLevel: 500
						}
					}}
				/>
			);
			const tabArea = renderer.create(
				component.instance().tabContentAreaConnected('generalSettings')
			);
			expect(tabArea.toJSON()).toMatchSnapshot();
		});

		it( 'Shows freeTrial tab', () => {
			const component = shallow(
				<ProSettings
					onSettingsSave={handler}
					proSettings={{
						connected: false,

					}}
				/>
			);
			const tabArea = renderer.create(
				component.instance().tabContentAreaNotConnected('freeTrial')
			);
			expect(tabArea.toJSON()).toMatchSnapshot();
		});

		it( 'Shows whatIs tab', () => {
			const component = shallow(
				<ProSettings
					onSettingsSave={handler}
					proSettings={{
						connected: false,

					}}
				/>
			);
			const tabArea = renderer.create(
				component.instance().tabContentAreaNotConnected('whatIs')
			);
			expect(tabArea.toJSON()).toMatchSnapshot();
		});
	});
});

