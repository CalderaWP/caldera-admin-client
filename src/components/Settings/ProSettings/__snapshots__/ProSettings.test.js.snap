// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProSettings component Matches snapshot with minimal props 1`] = `
<div
  className="caldera-forms-pro-settings"
>
  <div
    aria-orientation="vertical"
    className="components-tab-panel__tabs"
    onFocus={undefined}
    onKeyDown={[Function]}
    role="tablist"
  >
    <button
      aria-controls="0-apiKeys-view"
      aria-selected={true}
      className="pro-settings is-active"
      id="0-apiKeys"
      onClick={[Function]}
      role="tab"
      tabIndex={null}
    >
      Api Keys
    </button>
    <button
      aria-controls="0-generalSettings-view"
      aria-selected={false}
      className="pro-settings "
      id="0-generalSettings"
      onClick={[Function]}
      role="tab"
      tabIndex={-1}
    >
      Caldera Forms Pro Settings
    </button>
    <button
      aria-controls="0-formSettings-view"
      aria-selected={false}
      className="pro-form-settings "
      id="0-formSettings"
      onClick={[Function]}
      role="tab"
      tabIndex={-1}
    >
      Form Settings
    </button>
  </div>
  <div
    aria-labelledby="0-apiKeys"
    className="components-tab-panel__tab-content"
    id="0-apiKeys-view"
    role="tabpanel"
    tabIndex="0"
  />
</div>
`;
